lib/pbench/server/api/resources/__init__.py

```diff
@@ -0,0 +1,686 @@
+                access = new_data.get("access")  # normalized access policy
+                try:
+                    self._check_authorization(user, access)
+                except UnauthorizedAccess as e:
+                    self.logger.warning("{}", e)
+                    abort(HTTPStatus.FORBIDDEN, message="Not Authorized")
+            return method(new_data, request)
+        else:
+            return method(json_data, request)
+
+    def _get(self, json_data: JSON, request: Request) -> Response:
+        """
+        ABSTRACT METHOD: override in subclass to perform operation
+
+        Perform the requested GET operation, and handle any exceptions.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            request: Original incoming Request object
+
+        Returns:
+            Response to return to client
+        """
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} doesn't override abstract _get method"
+        )
+
+    def _post(self, json_data: JSON, request: Request) -> Response:
+        """
+        ABSTRACT METHOD: override in subclass to perform operation
+
+        Perform the requested POST operation, and handle any exceptions.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            request: Original incoming Request object
+
+        Returns:
+            Response to return to client
+        """
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} doesn't override abstract _post method"
+        )
+
+    def _put(self, json_data: JSON, request: Request) -> Response:
+        """
+        ABSTRACT METHOD: override in subclass to perform operation
+
+        Perform the requested PUT operation, and handle any exceptions.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            request: Original incoming Request object
+
+        Returns:
+            Response to return to client
+        """
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} doesn't override abstract _put method"
+        )
+
+    def _delete(self, json_data: JSON, request: Request) -> Response:
+        """
+        ABSTRACT METHOD: override in subclass to perform operation
+
+        Perform the requested DELETE operation, and handle any exceptions.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            request: Original incoming Request object
+
+        Returns:
+            Response to return to client
+        """
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} doesn't override abstract _delete method"
+        )
+
+    @Auth.token_auth.login_required(optional=True)
+    def get(self):
+        """
+        Handle an authenticated GET operation on the Resource
+        """
+        return self._dispatch(self._get, request)
```

Comment:
Where is `request` defined/declared/provided?
(Ditto for the other methods.)

Reply:
Flask magic ... this hasn't changed, it just stands out because I refactored the main entry points.